ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+        span.set_tag_str("langchain.request.provider", llm_provider)
+        span.set_tag_str("langchain.request.type", "chat_model")
+        span.set_tag_str("langchain.request.api_key", str(_extract_api_key(instance)))
+        for param, val in getattr(instance, "_identifying_params", {}).items():
+            if isinstance(val, dict):
+                for k, v in val.items():
+                    span.set_tag_str("langchain.request.%s.parameters.%s.%s" % (llm_provider, param, k), str(v))
+            else:
+                span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
+
+        chat_completions = func(*args, **kwargs)
+        if isinstance(instance, langchain.chat_models.ChatOpenAI):
+            _tag_openai_token_usage(span, chat_completions.llm_output)
+            integration.record_usage(span, chat_completions.llm_output)
+
+        for message_set_idx, message_set in enumerate(chat_completions.generations):
+            for idx, chat_completion in enumerate(message_set):
+                if integration.is_pc_sampled_span(span):
+                    span.set_tag_str(
+                        "langchain.response.completions.%d.%d.content" % (message_set_idx, idx),
+                        integration.trunc(chat_completion.text),
+                    )
+                span.set_tag_str(
+                    "langchain.response.completions.%d.%d.message_type" % (message_set_idx, idx),
+                    chat_completion.message.__class__.__name__,
+                )
+
+        if integration.is_pc_sampled_log(span):
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s" % (instance.__module__, instance.__class__.__name__),
+                attrs={
+                    "messages": [
+                        [
+                            {
+                                "content": message.content,
+                                "message_type": message.__class__.__name__,
+                            }
+                            for message in messages
+                        ]
+                        for messages in chat_messages
+                    ],
+                    "choices": [
+                        [
+                            {
+                                "content": message.text,
+                                "message_type": message.message.__class__.__name__,
+                            }
+                            for message in messages
+                        ]
+                        for messages in chat_completions.generations
+                    ],
+                },
+            )
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        integration.metric(span, "incr", "request.error", 1)
+        raise
+    finally:
+        span.finish()
+        integration.metric(span, "dist", "request.duration", span.duration_ns)
```

Comment:
There's some duplication in this section of multiple functions in this file. You might consider putting this logic into a function to make future changes easier.

Reply:
Ahh I did try doing this, but extracting the try/catch for the exception handling to a helper function breaks tracing - I'll add this as a follow up PR task.